Switches from empty boolean query to matchNoDocs

MaineC opened the pull request:
Closes #17981 

Caveat: Tests still running.

@cbuescher Input most welcome. I left one TODO with a question in the code.

cbuescher commented:
I wasn't aware that there were two MatchNoDocsQuery queries now, one in lucene itself (`org.apache.lucene.search`) and one in `org.elasticsearch.common.lucene.search`. The first is rewriting to the empty BooleanQuery, the second is able to print out this "reason" string. I'm not sure which one we are supposed to use in Queries#newMatchNoDocsQuery(). Maybe @jimferenczi has and opinion on this? We could use have Queries#newMatchNoDocsQuery(reason) to keep the reason for the missing matches, and provide another method for cases where the reason is not defined?

cbuescher commented on core/src/main/java/org/elasticsearch/index/search/MatchQuery.java:
```diff
@@ -286,6 +286,8 @@ protected final Query termQuery(MappedFieldType fieldType, Object value, boolean
 
     protected Query zeroTermsQuery() {
+        // TODO This is weird: DEFAULT_ZERO_TERMS_QUERY is a public static final field, so this can have only one value.
+        // why are we then having this if clause in here?
```
I think this is okay though, it checks if the current `zeroTermsQuery` is the same as the default, which is ZeroTermsQuery.NONE.

MaineC replied:
After re-checking what's going on I realised I missed the setter for zeroTermsQuery when reading this first. Removed the question.

MaineC commented:
> I'm not sure which one we are supposed to use in Queries#newMatchNoDocsQuery()

I'd like to try and only use the one requiring a reason string. As soon as we offer one that doesn't require the reason, my guess would be that the one with the reason isn't going to be used.

jimferenczi commented:
@cbuescher IMO we should always use the MatchNoDocsQuery(reason). There's always a reason why we use a MatchNoDocsQuery, most of the time it happens during the rewriting phase and the reason is  pretty clear.

MaineC commented:
Changed to MatchNoDocsQuery(reason) - sorry for the confusion, apparently IDE hadn't picked up the new class yet and I didn't look closely enough at the package of the imported class.

Added reason messages where appropriate, not entirely sure if all of them make sense as is, so double-checking appreciated.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/IdsQueryBuilder.java:
```diff
@@ -204,7 +204,7 @@ public String getWriteableName() {
         Query query;
         if (this.ids.isEmpty()) {
-             query = Queries.newMatchNoDocsQuery();
+             query = Queries.newMatchNoDocsQuery("Missing ids in " + this.getName());
```
nit: this will read "Missing ids in ids", might be a bit redundant. Not sure how to improve this, but maybe just quote it and add that it was a query?

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/ExistsQueryBuilder.java:
```diff
@@ -134,7 +134,7 @@ public static Query newFilter(QueryShardContext context, String fieldPattern) {
         if (fieldNamesFieldType == null) {
             // can only happen when no types exist, so no docs exist either
-            return Queries.newMatchNoDocsQuery();
+            return Queries.newMatchNoDocsQuery("Missing types in " + NAME);
```
same here, "Missing types in exists", maybe beter sth. like "Missing types in 'exists' query" or something like that.

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/MatchNoneQueryBuilder.java:
```diff
@@ -93,7 +93,7 @@ public static MatchNoneQueryBuilder fromXContent(QueryParseContext parseContext)
     @Override
     protected Query doToQuery(QueryShardContext context) throws IOException {
-        return Queries.newMatchNoDocsQuery();
+        return Queries.newMatchNoDocsQuery("User requested " + this.getName());
```
dito

cbuescher commented on core/src/main/java/org/elasticsearch/index/query/TermsQueryBuilder.java:
```diff
@@ -317,7 +317,7 @@ protected Query doToQuery(QueryShardContext context) throws IOException {
         }
         if (values == null || values.isEmpty()) {
-            return Queries.newMatchNoDocsQuery();
+            return Queries.newMatchNoDocsQuery("No terms supplied for " + getName());
```
dito

cbuescher commented on core/src/main/java/org/elasticsearch/index/search/MatchQuery.java:
```diff
@@ -286,7 +286,11 @@ protected final Query termQuery(MappedFieldType fieldType, Object value, boolean
     protected Query zeroTermsQuery() {
-        return zeroTermsQuery == DEFAULT_ZERO_TERMS_QUERY ? Queries.newMatchNoDocsQuery() : Queries.newMatchAllQuery();
+        if (zeroTermsQuery == DEFAULT_ZERO_TERMS_QUERY) {
+            return Queries.newMatchNoDocsQuery("No query supplied in match query");
```
Nit: The zeroTermsQuery was left (or intentionally set) to the default here. This happens according to the docs when 'no query terms are available, e.g. after analysis removed them'. So maybe just mention sth. like "Matching no documents because no terms present"?

cbuescher commented:
@MaineC thanks, I left a few comments about wording, other than that this LGTM
